### PR TITLE
Add additional trie node functionality

### PR DESCRIPTION
**Pull request by tcrain**
### What changes are proposed in this pull request?

Add additional trie node functionality.
- Store a typed value with each node
- A function to search an exact path
- A function to get the children of a path

### Why are the changes needed?

Additional functionality for new features.

### Does this PR introduce any user facing changes?

No

### CODE REVIEW

File: core/common/src/main/java/alluxio/conf/path/TrieNode.java [hunk 1]
```diff
@@ -140,10 +168,56 @@ public TrieNode deleteIf(String path, java.util.function.Function<TrieNode, Bool
...
     }
     return Iterators.concat(mChildren.values().stream().map(TrieNode::getCommonRoots).iterator());
   }
+
+  /**
+   * Get the terminal children of path (including path).
+   * @param path the path
+   * @return the terminal children
+   */
+  public Stream<TrieNode<V>> getChildren(String path) {
```
Comment (jenoudet) [inline core/common/src/main/java/alluxio/conf/path/TrieNode.java:183]:
```suggestion
  public Stream<TrieNode<V>> getLeafChildren(String path) {
```
Add `Leaf` or `Terminal` or some other identifier to the name of the function to indicate that it returns only terminal children.

File: core/common/src/main/java/alluxio/conf/path/TrieNode.java [hunk 2]
```diff
@@ -140,10 +168,56 @@ public TrieNode deleteIf(String path, java.util.function.Function<TrieNode, Bool
...
+
+  private Stream<TrieNode<V>> getChildrenInternal() {
+    return Stream.concat(Stream.of(this), mChildren.values().stream().flatMap(
+        TrieNode::getChildrenInternal));
+  }
+
+  /**
+   * Recursively removes all children.
+   */
+  public void clear() {
```
Comment (jenoudet) [inline core/common/src/main/java/alluxio/conf/path/TrieNode.java:217]: This method needs testing.

**Review by jenoudet — CHANGES_REQUESTED** [review]

**Review by jenoudet — APPROVED** [review]

**Review by jiacheliu3 — COMMENTED** [review]
Oops isn't this one very similar to https://github.com/Alluxio/alluxio/pull/15578/files#diff-be4d1e03f8d902b61762f11e492379d838046277460ec713b9e18505545a58c8
Could you take a look and see if there's anything that can be reused/merged?

**Comment by yuzhu** [comment]
alluxio-bot, merge this please

**Pull request merged**